pennylane/resource/resource.py

```diff
@@ -0,0 +1,164 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+The data class which will aggregate all the resource information from a quantum workflow.
+"""
+from typing import DefaultDict
+from collections import defaultdict
+
+
+class Resources:
+    r"""Contains attributes which store key resources (such as number of gates, number of wires, etc.)
+    tracked over a quantum workflow.
+
+    Args:
+        num_wires (int): number of qubits
+        num_gates (int): number of gates
+        gate_types (defaultdict(int)): dictionary with keys are operation names (str)
+            and values are the number of times the operation is repeated in the circuit (int)
+        depth (int): the depth of the circuit (max number of non-parallel operations)
+        shots (int): number of sampels to measure
+
+    Raises:
+        TypeError: If the attributes provided are not of the correct type.
+        AttributeError: If an attribute is set after initialization.
+
+    .. details::
+
+        The resources being tracked can be accessed and set as class attributes.
+        Additionally, the :code:`Resources` instance can be nicely displayed in the console.
+
+        **Example**
+
+        >>> r = Resources(num_wires=2, num_gates=2, gate_types={'Hadamard': 1, 'CNOT':1}, depth=2)
+        >>> print(r)
+        wires: 2
+        gates: 2
+        depth: 2
+        shots: 0
+        gate_types: {'Hadamard': 1, 'CNOT': 1}
+    """
+
+    def __init__(
+        self, num_wires=0, num_gates=0, gate_types=defaultdict(int), depth=0, shots=0
+    ):  # pylint: disable=too-many-arguments, dangerous-default-value
+        """Initialize a Resources instance and perform input type validation."""
+
+        if not (
+            all(isinstance(param, int) for param in [num_wires, num_gates, depth, shots])
+            and all(val > -1 for val in [num_wires, num_gates, depth, shots])
+            and isinstance(gate_types, DefaultDict)
+        ):
+            raise TypeError(
+                "Incorrect type of input, expected type int for num_wires, num_gates, depth and shots. "
+                f"Got {type(num_wires)}, {type(num_gates)}, {type(depth)}, {type(shots)} respectively. "
+                f"Expected type dict or DefaultDict for gate_types, got {type(gate_types)}."
+            )
+
+        self._num_wires = num_wires
+        self._num_gates = num_gates
+        self._gate_types = gate_types
+        self._depth = depth
+        self._shots = shots
+
+    @property
+    def num_wires(self):
+        """Number of wires in the quantum workflow."""
+        return self._num_wires
+
+    @num_wires.setter
+    def num_wires(self, new_val):
+        """Setting is NOT allowed"""
+        raise AttributeError(
+            f"{type(self)} object does not support assignment for 'num_wires' attribute"
+        )
```

Comment:
The idea is that a user should only ever want to read the values from a `Resources` object. This adds an extra layer of security in case anyone dynamically updates the values over a pipeline. One silly example would be as follows: 

```
Class custom_operation(Operation):
    def resources(self, *args, **kwargs):
        num_wires = 3
       num_gates, gate_types = get_custom_gates(*args, **kwargs)
        r = qml.Resources(num_wires, num_gates, gate_types)

        print(r.num_gates += 1)  # suppose the user wanted to print the number of gates + 1 
        return r                            # but accidentally types +=, then the value for num_gates is increased by 1 
```